addons/point_of_sale/models/pos_payment_method.py

```diff
@@ -0,0 +1,117 @@
+    is_cash_count = fields.Boolean(string='Cash')
+    cash_journal_id = fields.Many2one('account.journal',
+        string='Cash Journal',
+        domain=[('type', '=', 'cash')],
+        help='The payment method is of type cash. A cash statement will be automatically generated.')
+    split_transactions = fields.Boolean(
+        string='Split Transactions',
+        default=False,
+        help='If ticked, each payment will generate a separated journal item. Ticking that option will slow the closing of the PoS.')
+    session_ids = fields.Many2many('pos.session', string='Pos Sessions', help='Pos sessions that are using this payment method.')
+    config_ids = fields.Many2many('pos.config', string='Point of Sale Configurations')
+    company_id = fields.Many2one('res.company', string='Company', readonly=True, related='receivable_account_id.company_id')
+
+    @api.onchange('is_cash_count')
+    def _onchange_is_cash_count(self):
+        if not self.is_cash_count:
+            self.cash_journal_id = False
+
+    @api.multi
+    def write(self, vals):
+        active_sessions = self.mapped('session_ids').filtered(lambda session: session.state != 'closed')
+        if active_sessions:
+            raise UserError('Modifying payment methods used in an open session is not allowed.\n'
+                            'Open sessions: %s' % (' '.join(active_sessions.mapped('name')),))
+        return super(PosPaymentMethod, self).write(vals)
+
+    @api.model
+    def generate_bank_and_cash_payment_methods(self):
+        """ Used to create bank and cash payment methods.
+
+        The purpose of this function is to automatically generate
+        bank and cash payment methods when PoS module is installed
+        or when chart of accounts is installed in a company.
+
+        Thus, this is called when loading PoS data and in the inherited
+        function `_create_bank_journals` of account.chart.template model.
+        """
+        company = self.env.company
+        if not company.chart_template_id:
+            # Do not proceed if no chart of accounts is installed in the company.
+            return
+
+        cash_journals = self.env['account.journal'].search([('company_id', '=', company.id), ('type', '=', 'cash')])
+        if not cash_journals:
+            cash_journals = self._create_pos_cash_journal(company)
+
+        pos_receivable_account = company.default_pos_receivable_account_id
+        cash_payment_method = self._create_cash_payment_method(pos_receivable_account, cash_journals[0])
+        bank_payment_method = self._create_bank_payment_method(pos_receivable_account)
+
+        main_shop = self.env.ref('point_of_sale.pos_config_main', raise_if_not_found=False)
+        # pos_config_main might belong to other company. If so, do not add the payment methods.
+        if main_shop and main_shop.check_access_rights('write', raise_exception=False):
```

Comment:
not sure checking acces rights is relevent, because the method 'generate_bank_and_cash_payment_methods' should probably be called in 'sudo()'. The problem is that if you have checked multiple company lets assume company (1, 2) in the top widget, and add a chart of account on the company 2. You'll have the write access rights on the main shop which is in the company 1, and so probably replace the payment methods you've already set.

Maybe we just should check that the company of main shop = self.env.company. if it is called in sudo, maybe adding 'force_company' in the context is needed)

Reply:
This makes so much sense than what I implemented! You saved me from this very horrible code. :)